src/main/java/org/elasticsearch/plugins/repository/LocalPluginRepository.java

```diff
@@ -0,0 +1,433 @@
+        if ((source.getFileName() != null) && (source.getFileName().toString().endsWith(JAVA))) {
+            throw new PluginRepositoryException(name(), "Plugin contains source code, aborting installation");
+        }
+
+        boolean override = true;
+        boolean executable = false;
+
+        // Finds the destination of the file
+        Path destination = pluginsDir.resolve(name);
+        int depth = 0;
+
+        for (Path sub : source.toAbsolutePath()) {
+            // We search for 'bin/' or 'config/' in the 2 first path levels,
+            // if found we skip those directories
+            if (depth < (skippedLevels + 2)) {
+                if (sub.startsWith(BIN)) {
+                    destination = binDir.resolve(name);
+                    executable = true;
+                    continue;
+                } else if (sub.startsWith(CONFIG)) {
+                    destination = configDir.resolve(name);
+                    override = false;
+                    continue;
+                }
+            }
+
+            if (depth++ < skippedLevels) {
+                continue;
+            }
+            destination = destination.resolve(sub.toString());
+        }
+
+        if (Files.isDirectory(source)) {
+            try {
+                Files.createDirectories(destination);
+            } catch (FileAlreadyExistsException e) {
+                // Directory already exist, it's OK to ignore
+            }
+        } else {
+            // Do not override the file
+            if ((!override) && (Files.exists(destination))) {
+                if (FileSystemUtils.isSameFile(source, destination)) {
+                    // File is unchanged, do nothing
+                    return destination;
+                }
+                // We just append a suffix to the file
+                destination = destination.getParent().resolve(destination.toString().concat(".new"));
+            }
+
+            Files.copy(source, destination, StandardCopyOption.REPLACE_EXISTING);
+
+            // Makes the file executable
+            if ((executable) && (Files.getFileStore(destination).supportsFileAttributeView(PosixFileAttributeView.class))) {
+                if (Files.isRegularFile(destination)) {
+                    Files.setPosixFilePermissions(destination, Sets.newHashSet(PosixFilePermission.OWNER_EXECUTE,
+                            PosixFilePermission.GROUP_EXECUTE,
+                            PosixFilePermission.OTHERS_EXECUTE));
+                }
+            }
+        }
+
+        return destination;
+    }
+
+    @Override
+    public void remove(PluginDescriptor plugin) {
+        if (!Strings.hasText(plugin.name())) {
+            throw new PluginRepositoryException(name(), "Unable to install a plugin with a null or empty name");
+        }
+
+        validatePluginName(plugin.name());
+
+        try {
+            // Remove the plugin's files in 'bin/' and 'plugins/' directories
```

Comment:
Global note about english: sounds like my spell checker prefers `the plugin files` than the `plugin's files`. I know. It confuses me as well as we learned the later form at school :)